### PR TITLE
[Facts of the day] #7384 Add additional display of fact of the day by habit tag acquired by the user 

### PR DESCRIPTION
**Pull request by kryzanivska-nastya**
## Issue Link 📋
[#7384](https://github.com/ita-social-projects/GreenCity/issues/7384)

##
**Summary Of Changes 🔥**

**_Developed the UI mechanism for displaying the additional "Fact of the Day" on the "My Account" page, ensuring that it updates daily._** 

-  Added a new card to display the habitFactOfTheDay content in the UI when available.

- Introduced habitFactOfTheDay for managing habit-related facts.

- **ProfileService**: Added getFactsOfTheDayByTags() to fetch habit facts.

- **LocalStorageService**: Added methods to save, retrieve, and clear habit facts with a 24-hour validity check.

- Logic to verify if more than 24 hours have passed since the last fetch and to update the habit facts accordingly.

- Replaced individual subscriptions with a Subject<void> for cleaner subscription handling.

Знімок екрана 2024-09-15 о 11 45 51

### CODE REVIEW

File: src/app/main/component/user/components/profile/profile-cards/profile-cards.component.html [hunk 1]
```diff
@@ -6,4 +6,12 @@
       <img src="assets/img/wave_shape.png" alt="wave-background" />
     </div>
   </div>
+
+  <div class="card habit-card" *ngIf="habitFactOfTheDay">
+    <p class="cart-title">{{ 'profile.fact-of-the-day' | translate }}</p>
+    <p class="card-description">{{ habitFactOfTheDay?.content }}</p>
+    <div class="shape-img">
+      <img src="assets/img/wave_shape.png" alt="wave-background" />
```
Comment (jaworovcki) [inline src/app/main/component/user/components/profile/profile-cards/profile-cards.component.html:14]: Is it okay to hardcode the src in html directly?

File: src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts [hunk 2]
```diff
@@ -23,8 +26,27 @@ export class ProfileCardsComponent implements OnInit, OnDestroy {
...
       this.getFactOfTheDay();
+      this.checkAndUpdateHabitFact();
     });
+
+    this.dailyUpdateSubscription = timer(0, 24 * 60 * 60 * 1000).subscribe(() => {
+      this.checkAndUpdateHabitFact();
+    });
+  }
+
+  loadFactsFromLocalStorage() {
```
Comment (jaworovcki) [inline src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts:39]: I presume that this logic should be in the service

File: src/app/main/component/user/components/profile/profile-cards/profile-cards.component.html (hunk 1 above)
Comment (kryzanivska-nastya) [reply src/app/main/component/user/components/profile/profile-cards/profile-cards.component.html:14]: It was there for the first fact, so I decided to go with that

File: src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts (hunk 2 above)
Comment (kryzanivska-nastya) [reply src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts:39]: Yes, you are right, I planned to do that and redo the tests

File: src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts [hunk 3]
```diff
@@ -23,8 +26,27 @@ export class ProfileCardsComponent implements OnInit, OnDestroy {
...
+    const savedHabitFact = localStorage.getItem('habitFactOfTheDay');
+    const lastHabitFetchTime = localStorage.getItem('lastHabitFactFetchTime');
+    const currentTime = Date.now();
+    const oneDay = 24 * 60 * 60 * 1000;
```
Comment (jaworovcki) [inline src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts:43]: repeated more then one time use of this expression, would it be better to create a global const from it?

File: src/app/main/component/user/components/profile/profile-cards/profile-cards.component.html (hunk 1 above)
Comment (Lokankara) [inline src/app/main/component/user/components/profile/profile-cards/profile-cards.component.html:14]: You can put the image path in the folder`src/app/main/image-pathes/profile-` and import the path from there
Use ngSrc (NgOptimizedImage directive) to improve performance

File: src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts [hunk 4]
```diff
@@ -1,19 +1,20 @@
 import { Component, OnInit, OnDestroy } from '@angular/core';
-import { catchError } from 'rxjs/operators';
-import { Subscription } from 'rxjs';
+import { catchError, takeUntil } from 'rxjs/operators';
+import { Subject, Subscription, of, timer } from 'rxjs';
```
Comment (Lokankara) [inline src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts]: Remove this unused import of 'Subscription'.

File: src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts [hunk 5]
```diff
@@ -22,27 +23,83 @@ export class ProfileCardsComponent implements OnInit, OnDestroy {
...
+      });
+  }
+
+  updateHabitFactIfNeeded(): void {
+    const lastHabitFetchTime = localStorage.getItem('lastHabitFactFetchTime');
+    const currentTime = Date.now();
+    const oneDay = this.profileService.oneDayInMillis;
+
+    // Only update habit fact if the last fetch was over 24 hours ago
+    if (!lastHabitFetchTime || currentTime - Number(lastHabitFetchTime) > oneDay) {
```
Comment (Lokankara) [inline src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts]: You can improved the logic by separating time difference calculation from the condition check
Please remove the comment

File: src/app/main/component/user/components/profile/profile-service/profile.service.ts [hunk 6]
```diff
@@ -14,6 +14,7 @@ import { Patterns } from 'src/assets/patterns/patterns';
   providedIn: 'root'
```
Comment (Lokankara) [inline src/app/main/component/user/components/profile/profile-service/profile.service.ts]: Please remove the alias `FactOfTheDay as FactOfTheDay` from line 4.

File: src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts [hunk 7]
```diff
@@ -22,27 +23,83 @@ export class ProfileCardsComponent implements OnInit, OnDestroy {
...
+      .pipe(takeUntil(this.destroy$))
+      .subscribe(() => this.checkAndUpdateHabitFact());
+  }
+
+  loadHabitFactFromLocalStorage(): void {
+    this.habitFactOfTheDay = this.profileService.getHabitFactFromLocalStorage();
   }
 
   getFactOfTheDay(): void {
-    this.profileSubscription = this.profileService
```
Comment (Lokankara) [inline src/app/main/component/user/components/profile/profile-cards/profile-cards.component.ts:31]: you removed this.profileSubscription, but did not remove ProfileCardsComponent.spec
please update `should unsubscribe from subscriptions on component destroy`
`component.languageSubscription`
`component.profileSubscription`

File: src/app/main/component/user/components/profile/profile-service/profile.service.ts [hunk 8]
```diff
@@ -46,6 +47,32 @@ export class ProfileService {
...
+    const lastHabitFetchTime = localStorage.getItem('lastHabitFactFetchTime');
+    const currentTime = Date.now();
+
+    if (savedHabitFact && lastHabitFetchTime && currentTime - Number(lastHabitFetchTime) < this.oneDayInMillis) {
+      return JSON.parse(savedHabitFact);
+    }
+    return null;
+  }
+
+  saveHabitFactToLocalStorage(fact: FactOfTheDay, currentTime: number): void {
```
Comment (Lokankara) [inline src/app/main/component/user/components/profile/profile-service/profile.service.ts]: The main purpose of the `ProfileService` is to fetch data from the server
If you need `SaveHabitFactToLocalStorage` or `clearHabitFactFromLocalStorage`, you can create them in LocalStorageService

File: src/app/main/component/user/components/profile/profile-service/profile.service.ts [hunk 9]
```diff
@@ -46,6 +47,32 @@ export class ProfileService {
     return this.http.get<FactOfTheDay>(`${mainLink}fact-of-the-day/random?lang=${currentLang}`);
   }
 
+  getHabitFactFromLocalStorage(): FactOfTheDay | null {
+    const savedHabitFact = localStorage.getItem('habitFactOfTheDay');
+    const lastHabitFetchTime = localStorage.getItem('lastHabitFactFetchTime');
+    const currentTime = Date.now();
+
+    if (savedHabitFact && lastHabitFetchTime && currentTime - Number(lastHabitFetchTime) < this.oneDayInMillis) {
```
Comment (Lokankara) [inline src/app/main/component/user/components/profile/profile-service/profile.service.ts]: You can simplify the condition and break it down into smaller, easier to read parts

**Review by hnativlyubomyr — APPROVED** [review]

**Review by Lokankara — APPROVED** [review]

**Pull request merged**